walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // if the request meeting is longer than the duration of one day
+    // return an empty list
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // if there is no event on the calendar currently
+    // return whole day as possible time slot
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // if no mandatory attendees or optional attendees, return whole day
+    if (mandatoryAttendees.size() == 0 && optionalAttendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // get unavailable time of mandatory attendees
+    List<TimeRange> unavailableTime = getUnavailableTime(events, mandatoryAttendees);
+    // merge available intervals with overlaps
+    List<TimeRange> mergedUnavailableTime = mergeUnavailableTime(unavailableTime);
+    // get available time slot base on merged unavailable time
+    List<TimeRange> res = getAvailableTime(mergedUnavailableTime, request.getDuration());
+    return res;
+  }
+
+  private List<TimeRange> getUnavailableTime(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> res = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      for (String attendee : attendees) {
+        // if there is a least on mandatory attendee for the requested event need to attend this event
+        // we mark the duration of this event as unavailable
+        if (mandatoryAttendees.contains(attendee)) {
+          res.add(e.getWhen());
+          break;
+        }
+      }
+    }
+    return res;
+  }
+
+  private List<TimeRange> mergeUnavailableTime(List<TimeRange> timeRanges) {
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> res = new ArrayList<>();
+
+    for (TimeRange t : timeRanges) {
+      // if the result list is empty, we can add it to the result list directly
+      if (res.isEmpty()) {
+        res.add(t);
+      } else {
+        TimeRange latestTimeRange = res.get(res.size() - 1);
+
+        // if the latest time slot that has been processed doesn't overlap with the time slot currently being processed
+        if (!t.overlaps(latestTimeRange)) {
```

Comment:
I would move this comment after the if check so it's clear that it's referring to the case where there is no overlap.

Maybe reword to something like: `If the previous time slot doesn't overlap with the current time slot, we can add it to the result list directly`